error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/NullRules.java

```diff
@@ -43,22 +47,50 @@ boolean after(@Nullable Object object) {
     }
   }
 
-  /** Prefer {@link Objects#requireNonNullElse(Object, Object)} over the Guava alternative. */
-  // XXX: This rule is not valid in case `second` is `@Nullable`: in that case the Guava variant
-  // will return `null`, while the JDK variant will throw an NPE.
+  /**
+   * Prefer {@link Objects#requireNonNullElse(Object, Object)} over the Guava and Optional
+   * alternatives.
+   */
+  // XXX: This rule is not valid in case `second` is `@Nullable`: in that case the Guava and
+  // Optional variants
+  // will return `null`, while the Objects.requireNonNullElse variant will throw an NPE.
   static final class RequireNonNullElse<T> {
     @BeforeTemplate
     T before(T first, T second) {
       return MoreObjects.firstNonNull(first, second);
     }
 
+    @BeforeTemplate
+    T beforeOptional(T first, T second) {
+      return Optional.ofNullable(first).orElse(second);
+    }
+
     @AfterTemplate
     @UseImportPolicy(STATIC_IMPORT_ALWAYS)
     T after(T first, T second) {
       return requireNonNullElse(first, second);
     }
   }
 
+  /**
+   * Prefer {@link Objects#requireNonNullElseGet(Object, Supplier)} over {@link Optional}{@code
+   * .ofNullable(Object).orElseGet(Supplier)}
+   */
+  // XXX: This rule is not valid in case `supplier` may return `null`: Optional will return `null`,
+  // while the `requireNonNullElseGet` replacement will throw an NPE.
+  static final class RequireNonNullElseGet<T, S extends Supplier<T>> {
```

Comment:
This kind of subtype matching is done automatically by Refaster. That said, there's indeed work to be done here: we should introduce `S extends T`, so that we can match ` Supplier<S>`.